FastSimulation/SimplifiedGeometryPropagator/plugins/FastTrackDeDxProducer.cc

```diff
@@ -0,0 +1,218 @@
+	
+    DeDxTools::makeCalibrationMap(m_calibrationPath, *tkGeom, calibGains, m_off);
+  }
+
+  m_estimator->beginRun(run, iSetup);
+}
+
+
+
+void FastTrackDeDxProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+
+  auto trackDeDxEstimateAssociation = std::make_unique<ValueMap<DeDxData>>();  
+  ValueMap<DeDxData>::Filler filler(*trackDeDxEstimateAssociation);
+
+  edm::Handle<reco::TrackCollection> trackCollectionHandle;
+  iEvent.getByToken(m_tracksTag,trackCollectionHandle);
+  std::vector<DeDxData> dedxEstimate( trackCollectionHandle->size() );
+
+
+
+  for(unsigned int j=0;j<trackCollectionHandle->size();j++){            
+    const reco::TrackRef track = reco::TrackRef( trackCollectionHandle.product(), j );
+    
+    int NClusterSaturating = 0; 
+      DeDxHitCollection dedxHits;
+      
+      auto const & trajParams = track->extra()->trajParams();
+      assert(trajParams.size()==track->recHitsSize());
+      
+      //std::cout << "track->recHitsSize() = " << track->recHitsSize() << std::endl;
+      
+      auto hb = track->recHitsBegin();
+      dedxHits.reserve(track->recHitsSize()/2);
+      for(unsigned int h=0;h<track->recHitsSize();h++){
+	const FastTrackerRecHit recHit = static_cast< const FastTrackerRecHit & >(*(*(hb+h)));
+      //FastTrackerRecHit recHit = *(hb+h);
+	if(!recHit.isValid()) continue;
+	
+	auto trackDirection = trajParams[h].direction();         
+	float cosine = trackDirection.z()/trackDirection.mag();
+	processHit(recHit, track->p(), cosine, dedxHits, NClusterSaturating);
+       }
+  
+       sort(dedxHits.begin(),dedxHits.end(),less<DeDxHit>());   
+       std::pair<float,float> val_and_error = m_estimator->dedx(dedxHits);
+       
+       //WARNING: Since the dEdX Error is not properly computed for the moment
+       //It was decided to store the number of saturating cluster in that dataformat
+       val_and_error.second = NClusterSaturating; 
+       dedxEstimate[j] = DeDxData(val_and_error.first, val_and_error.second, dedxHits.size() );
+    }
+
+
+  filler.insert(trackCollectionHandle, dedxEstimate.begin(), dedxEstimate.end());
+  // fill the association map and put it into the event
+  filler.fill();
+  iEvent.put(std::move(trackDeDxEstimateAssociation));
+}
+
+
+void FastTrackDeDxProducer::processHit(const FastTrackerRecHit &recHit, float trackMomentum, float& cosine, reco::DeDxHitCollection& dedxHits, int& NClusterSaturating){
+
+
+  auto const & thit = static_cast<BaseTrackerRecHit const&>(recHit);
+  //const TrackerSingleRecHit thit = static_cast<TrackerSingleRecHit const&>(recHit);//playing now
+
+
+  if(!thit.isValid())return;
+
+  if(recHit.isPixel()){
+    //std::cout << "we got pixels" << std::endl;
+    if(!usePixel) return;
+
+    auto& detUnit     = *(recHit.detUnit());
+    float pathLen     = detUnit.surface().bounds().thickness()/fabs(cosine);
+    if (nothick) pathLen = 1.0;
+    // float chargeAbs   = clus.pixelCluster().charge();
```

Comment:
remove the commented out line if not needed, or add a comment which explains why you want to keep in the code